### PR TITLE
Add leave one out split to evaluation module.

### PR DESCRIPTION
**Pull request by markdouthwaite**
# Description

This PR adds a train/test split function (`leave_k_out_split`) that implements a 'leave `k` out' splitting policy. This involves randomly sampling exactly `k` interactions for each (eligible) user and placing these into the test set, while placing all other interactions in the training set. This is typically better suited than 'traditional' fully randomized train/test split functions for some evaluation tasks. The implementation here optionally enables the user to set `train_only_size` to ensure a given percentage of users appear _only_ in the training matrix.

## Changes

* **Added** - `implicit.evaluation.leave_k_out_split` splitting function.
* **Added** - Tests to cover the above function.
* **Updated** - `.gitignore` to include `.idea/*` dir.
* **Updated** - `implicit.evaluation` to use `isort`, `black`.

## Non-functional info

On an instance with 8 vCPUs and 16GB RAM, this implementation splits the Movielens 20m dataset in <5s for `K=1`. When using `K=1` with 80% of users reserved for training only, this drops to ~2.5s. In comparison, the existing `train_test_split` implementation splits the given dataset in ~1.5s.

## Notes

* The `leave_k_out_split` implementation is fully vectorized only for `K=1`. For `K>1`, an expensive list comprehension is used.

### CODE REVIEW

**Comment by markdouthwaite** [comment]
Could do with adding tests for `K>2` too. 🤔

**Review by ita9naiwa — COMMENTED** [review]
I think the code is fine, but one of the main reasons for doing leave-k-out evaluation in recommending tasks is for faster evaluation.

However, I think this leave-one-out is not faster enough than the usual 80:20 train:test split to compensate at the expense of [lower performance prediction accuracy](https://arxiv.org/abs/1905.01395)

**Comment by markdouthwaite** [comment]
Thanks for the review @ita9naiwa! Is it worth me spending a bit of time accelerating the implementation? 

As an aside, I saw this PR as a stepping stone towards adding the common evaluation approach of using leave-one-out to draw a single positive item sample, append `n` negative/random item samples, rank the samples and compute nDCG etc. (similar to the protocol described in [here](https://people.engr.tamu.edu/huangrh/Spring16/papers_course/matrix_factorization.pdf) and [here](https://arxiv.org/pdf/1708.05031.pdf)). Would that be useful at all?

**Comment by ita9naiwa** [comment]
>As an aside, I saw this PR as a stepping stone towards adding the common evaluation approach of using leave-one-out to draw a single positive item sample, append n negative/random item samples, rank the samples and compute nDCG etc. (similar to the protocol described in here and here). Would that be useful at all?


I agree with you. I think it should support sampling a few(100 in NeuMF paper) negative samples, and k(many cases smaller than 10) positive items.

than ranking among those sampled items will take less time than that of the current implementation.

I think this can be done with `recommender_base.rank_items` method, but (i don't know but) there will be a faster way to do this.

**Comment by markdouthwaite** [comment]
Okay, great. I'll optimise this PR to improve performance, and then look at raising a second PR with the evaluation protocol implementation.

File: .gitignore [hunk 1]
```diff
@@ -73,3 +73,4 @@ target/
 .ipynb_checkpoints
 
 .vscode/
+.idea/
```
Comment (ita9naiwa) [inline .gitignore:76]: I think this might be removed.

**Review by ita9naiwa — COMMENTED** [review]
`.cpp` files are preferred not be included in the code repository (they're generated automatically with `.pyx` files)

I think other parts are good...

File: .gitignore (hunk 1 above)
Comment (markdouthwaite) [reply .gitignore:76]: I added this as I saw the `.vscode` settings were included and thought it might be worthwhile adding `.idea` (PyCharm) settings too. Happy to remove it though, if you'd prefer 🙂

**Comment by markdouthwaite** [comment]
Hey @ita9naiwa! Thanks - I've dropped the auto-generated `.cpp` files as suggested.  👍

File: .gitignore (hunk 1 above)
Comment (ita9naiwa) [reply .gitignore:76]: Sorry i didnt know that. It will be fine i think.

**Review by ita9naiwa — APPROVED** [review]
Seems fine.

this functionality is required to compare some works which report their performances using leave-k evaluations.
